xdsl/tools/tblgen_to_py.py

```diff
@@ -0,0 +1,566 @@
+from sys import stdin
+from typing import Any
+
+
+@dataclass
+class TblgenRecord:
+    js: Any
+
+    def __getitem__(self, i: str) -> Any:
+        return self.js[i]
+
+    @property
+    def name(self) -> str:
+        return self["!name"]
+
+    @property
+    def summary(self) -> str:
+        return self["summary"]
+
+    @property
+    def superclasses(self) -> set[str]:
+        return set(self["!superclasses"])
+
+    @property
+    def dialect(self) -> str:
+        return self["dialect"]["def"]
+
+
+class TblgenOp(TblgenRecord):
+    @property
+    def dialect(self) -> str:
+        return self["opDialect"]["def"]
+
+    @property
+    def opName(self) -> str:
+        return self["opName"]
+
+    @property
+    def assemblyFormat(self) -> str | None:
+        if "assemblyFormat" in self.js:
+            assembly = self["assemblyFormat"]
+            if isinstance(assembly, str):
+                return assembly
+        return None
+
+    def _dag_to_py(self, dag: str) -> tuple[tuple[str, str], ...]:
+        args = self[dag]["args"]
+        return tuple((a["def"], n) for a, n in args)
+
+    @property
+    def arguments(self) -> tuple[tuple[str, str], ...]:
+        return self._dag_to_py("arguments")
+
+    @property
+    def results(self) -> tuple[tuple[str, str], ...]:
+        return self._dag_to_py("results")
+
+    @property
+    def regions(self) -> tuple[tuple[str, str], ...]:
+        return self._dag_to_py("regions")
+
+    @property
+    def successors(self) -> tuple[tuple[str, str], ...]:
+        return self._dag_to_py("successors")
+
+
+class TblgenType(TblgenRecord):
+    @property
+    def typeName(self) -> str:
+        return self["typeName"]
+
+
+class TblgenAttr(TblgenRecord):
+    @property
+    def attrName(self) -> str:
+        return self["attrName"]
+
+
+@dataclass
+class TblgenLoader:
+    """
+    Class for converting a json generated by `llvm-tblgen --dump-json` to python
+    code which builds an xDSL dialect. Must be initialised with json represented
+    as a python object. The generated code is stored as strings in the
+    `attributes` and `operations` dictionaries.
+    """
+
+    js: Any
+
+    attributes: dict[str, str] = field(default_factory=dict)
+    operations: dict[str, str] = field(default_factory=dict)
+    used_records: set[str] = field(default_factory=set)
+    anon_counter: int = field(default_factory=int)
+
+    def _get_op(self, name: str) -> TblgenOp:
+        self.used_records.add(name)
+        return TblgenOp(self.js[name])
+
+    def _get_type(self, name: str) -> TblgenType:
+        self.used_records.add(name)
+        return TblgenType(self.js[name])
+
+    def _get_attr(self, name: str) -> TblgenAttr:
+        self.used_records.add(name)
+        return TblgenAttr(self.js[name])
+
+    def _get_record(self, name: str) -> TblgenRecord:
+        self.used_records.add(name)
+        return TblgenRecord(self.js[name])
+
+    def load_dialect(self, dialect_name: str):
```

Comment:
I would rename these functions (and the class name), since now they generate a python xDSL dialect rather than loading one.

Reply:
I went with `import` rather than `load` but am not really sure if this is better

Reply:
`generate` could also work maybe?